### PR TITLE
Defer reroute when nodes join

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the master eagerly reroutes the cluster as part of processing node joins.
However, it is not necessary to do this reroute straight away, and it is
sometimes preferable to defer it until later. For instance, when the master
wins its election it processes joins and performs a reroute, but it would be
better to defer the reroute until after the master has become properly
established.

This change defers this reroute into a separate task.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
@ywelsch I managed to synthesise a failure by changing some of the `randomBoolean()`s and `rarely()`s in `IndicesClusterStateServiceRandomUpdatesTests#testRandomClusterStateUpdates`, but the probability of hitting it seemed very low otherwise.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 1]
```diff
@@ -141,15 +142,17 @@
     public Coordinator(String nodeName, Settings settings, ClusterSettings clusterSettings, TransportService transportService,
```
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: I think we're at the point when the number of Coordinator constructor parameters are unmanageable. Can we possibly add a JavaDoc to the constructor describing how specific dependency is used by Coordinator?

File: server/src/test/java/org/elasticsearch/cluster/SimpleDataNodesIT.java [hunk 2]
```diff
@@ -66,13 +74,62 @@ public void testDataNodes() throws Exception {
...
+        assertThat(healthResponse1.getActiveShards(), equalTo(0));
+
+        internalCluster().startNode(Settings.builder().put(Node.NODE_DATA_SETTING.getKey(), true).build());
+
+        assertThat(client().admin().cluster().prepareHealth()
+                .setWaitForEvents(Priority.LANGUID).setWaitForNodes("2").setWaitForGreenStatus().execute().actionGet().isTimedOut(),
+            equalTo(false));
+
+        final AtomicBoolean stopRerouting = new AtomicBoolean();
+        final Thread rerouteThread = new Thread(() -> {
```
Comment (andrershov) [inline server/src/test/java/org/elasticsearch/cluster/SimpleDataNodesIT.java]: I'm not sure I understand what reroute thread is doing here. I mean I understand that it continuously performs reroutes, but why is it needed in the test? Can you add the comment, please?

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: I know what you mean, but most of them are of very specific types. I added docs for the ones whose types don't make their meaning so clear in d9dbf2d (including the one added here).

File: server/src/test/java/org/elasticsearch/cluster/SimpleDataNodesIT.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/SimpleDataNodesIT.java]: I found a simpler way to test the same thing now that I understand what's going on a bit better. See 563ea02.

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java [hunk 3]
```diff
@@ -1243,7 +1243,8 @@ public void start(ClusterState initialState) {
                 allocationService, masterService, () -> persistedState,
                 hostsResolver -> testClusterNodes.nodes.values().stream().filter(n -> n.node.isMasterNode())
                     .map(n -> n.node.getAddress()).collect(Collectors.toList()),
-                clusterService.getClusterApplierService(), Collections.emptyList(), random());
+                clusterService.getClusterApplierService(), Collections.emptyList(), random(),
+                s -> {});
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java]: perhaps plug in in the actual RoutingService here. This test cares about shards and realistic mocking, so I'm worried that a lot of effort will be spend in the future here to figure out why shards are not allocated.

**Review by ywelsch — APPROVED** [review]
One nit, looking good o.w.

**Pull request merged**